src/types/params.ts

```diff
@@ -96,18 +96,33 @@ export type ExtractRouteParamTypes<TRoute> = TRoute extends {
   : Record<string, unknown>
 
 /**
- * Extracts combined types of path and query parameters for a given route, creating a unified parameter object.
- * This parameter object type represents the expected type when accessing params from router.route or useRoute.
+ * Does everything that ExtractRouteParamTypes does, but takes into consideration optional properties.
+ * Differs from ExtractRouteParamTypes in that it also reads the string `value` from WithParams to determine what should be optional.
+ * @template TRoute - The route type from which to extract and merge parameter types.
+ * @returns A record of parameter names to their respective types, extracted and merged from both path and query parameters.
+ */
+export type ExtractRouteParamTypesOptionalReading<TRoute> = TRoute extends {
+  host: { params: infer HostParams extends Record<string, Param> },
+  path: { params: infer PathParams extends Record<string, Param> },
+  query: { params: infer QueryParams extends Record<string, Param> },
+  hash: { params: infer HashParams extends Record<string, Param> },
+}
+  ? Identity<MakeOptional<ExtractParamTypesOptionalReading<HostParams> & ExtractParamTypesOptionalReading<PathParams> & ExtractParamTypesOptionalReading<QueryParams> & ExtractParamTypesOptionalReading<HashParams>>>
+  : Record<string, unknown>
+
+/**
+ * Does everything that ExtractRouteParamTypes does, but takes into consideration optional properties.
+ * Differs from ExtractRouteParamTypesOptionalReading in that optional params with defaults will remain optional.
  * @template TRoute - The route type from which to extract and merge parameter types.
  * @returns A record of parameter names to their respective types, extracted and merged from both path and query parameters.
  */
-export type ExtractRouteParamTypesWithOptional<TRoute> = TRoute extends {
+export type ExtractRouteParamTypesOptionalWriting<TRoute> = TRoute extends {
   host: { params: infer HostParams extends Record<string, Param> },
   path: { params: infer PathParams extends Record<string, Param> },
   query: { params: infer QueryParams extends Record<string, Param> },
   hash: { params: infer HashParams extends Record<string, Param> },
 }
-  ? ExtractParamTypesWithOptional<HostParams & PathParams & QueryParams & HashParams>
+  ? Identity<MakeOptional<ExtractParamTypesOptionalWriting<HostParams> & ExtractParamTypesOptionalWriting<PathParams> & ExtractParamTypesOptionalWriting<QueryParams> & ExtractParamTypesOptionalWriting<HashParams>>>
```

Comment:
Why do we need to call the extract type for each type of param now rather than passing all the params in as a union like we were before?

Reply:
in an upcoming PR I'm adjusting the `ExtractParam` types to also take the `string` part to determine what's optional as part of [this change](https://github.com/kitbagjs/router/pull/464). When that happens, this is how I want to call them so I left that in for now.